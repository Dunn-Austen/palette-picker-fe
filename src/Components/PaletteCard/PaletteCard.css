.palette-card {
  align-items: center;
  background-color: white;
  border: 1px solid black;
  display: flex;
  flex-direction: row;
  height: 10vh;
  justify-content: space-between;
  margin: 10px 0px;
  width: 90%;
}

.palette-container {
  border-top: 1px solid black;
  border-right: 1px solid black;
  border-bottom: 1px solid black;
  display: flex;
  flex-direction: row;
  height: 100%;
  width: 65%;
}

.btn-container {
  align-items: center;
  display: flex;
  flex-direction: column;
  height: 100%;
  justify-content: space-around;
  width: 35%;
}

.palette-slice {
  height: 100%;
  width: 20%;
}
