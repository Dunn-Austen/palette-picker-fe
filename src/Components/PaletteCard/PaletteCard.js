import React, { useState } from 'react';
import './PaletteCard.css';
import PropTypes from 'prop-types';
import { patchPalette, deletePalette, fetchAllPalettes, fetchAllProjects } from '../../apiCalls';

const PaletteCard = ({updateProjects, projects, id, title, project_id, color_1_id, color_2_id, color_3_id, color_4_id, color_5_id}) => {

  let [editStatus, setEditStatus] = useState(false);
  let [currentPalette, setCurrentPalette] = useState({id, title, color_1_id, color_2_id, color_3_id, color_4_id, color_5_id, project_id});

  const changeColor = (event) => {
    let newPalette = {...currentPalette};
    let targetedValue = event.target.value;
    newPalette[targetedValue] = '#'+Math.floor(Math.random()*16777215).toString(16);

    setCurrentPalette(newPalette);
  }

  const updatePalette = async (palette) => {
    let newProjects = [...projects];
    await patchPalette(palette)
      .then(returnedPalette => {
        let foundProject = newProjects.find(project => {
          return project.id === returnedPalette.project_id
        })
        let foundPalette = foundProject.palettes.find(palette => palette.id === returnedPalette.id);
        foundPalette.color_1_id = returnedPalette.color_1_id;
        foundPalette.color_2_id = returnedPalette.color_2_id;
        foundPalette.color_3_id = returnedPalette.color_3_id;
        foundPalette.color_4_id = returnedPalette.color_4_id;
        foundPalette.color_5_id = returnedPalette.color_5_id;

        updateProjects(newProjects);
        setEditStatus(!editStatus)
      });
  }

  const removePalette = async (event) => {
    const id = event.target.dataset.id;
    let temporaryProjects;

    await deletePalette(id);
    await fetchAllProjects()
      .then(projectsData => {
        temporaryProjects = projectsData.projects.map(project => {
          return {
            id: project.id,
            title: project.title,
            palettes: []
          }
        });
      });
    await fetchAllPalettes()
      .then(palettesData => {
          palettesData.palettes.forEach(palette => {
            temporaryProjects.find(project => palette.project_id === project.id).palettes.push(palette)
          })
          updateProjects(temporaryProjects)
      })
  }

  return (
    <section className='palette-card'>
      <section className='palette-container'>
        <div className='palette-slice' style={{backgroundColor: currentPalette.color_1_id}}>
          {editStatus &&
            <button className='change-color' value='color_1_id' onClick={changeColor}>
              Change
            </button>
          }
        </div>
        <div className='palette-slice' style={{backgroundColor: currentPalette.color_2_id}}>
          {editStatus &&
            <button className='change-color' value='color_2_id' onClick={changeColor}>
              Change
            </button>
          }
        </div>
        <div className='palette-slice' style={{backgroundColor: currentPalette.color_3_id}}>
          {editStatus &&
            <button className='change-color' value='color_3_id' onClick={changeColor}>
              Change
            </button>
          }
        </div>
        <div className='palette-slice' style={{backgroundColor: currentPalette.color_4_id}}>
          {editStatus &&
            <button className='change-color' value='color_4_id' onClick={changeColor}>
              Change
            </button>
          }
        </div>
        <div className='palette-slice' style={{backgroundColor: currentPalette.color_5_id}}>
        {editStatus &&
          <button className='change-color' value='color_5_id' onClick={changeColor}>
            Change
          </button>
        }
        </div>
      </section>
      <section className='btn-container'>
        {!editStatus &&
          <button className='edit-palette' onClick={() => setEditStatus(!editStatus)}>Edit</button>
        }
        {editStatus &&
          <button className='save-palette' onClick={() => updatePalette(currentPalette)}>Save</button>
        }
        <button data-id={id} className='delete-palette' onClick={removePalette}>Delete</button>
      </section>
    </section>
  )
}

export default PaletteCard;
