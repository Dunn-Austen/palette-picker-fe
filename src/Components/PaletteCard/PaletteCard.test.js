import React from 'react';
import { shallow } from 'enzyme';
import PaletteCard from './PaletteCard';

describe('PaletteCard', () => {
  it('should match the snapshot', () => {
    const mockUpdateProjects = jest.fn();
    const mockProjects = [
      {
        id: 1,
        title: 'Kitchen'
      }
    ];
    const mockId = 1;
    const mockTitle = 'wall';
    const mockProject_id = 23;
    const mockColor_1_id = '#bbbd33';
    const mockColor_2_id = '#bbbd33';
    const mockColor_3_id = '#bbbd33';
    const mockColor_4_id = '#bbbd33';
    const mockColor_5_id = '#bbbd33';
    const wrapper = shallow(<PaletteCard 
      updateProjects={mockUpdateProjects}
      projects={mockProjects}
      id={mockId}
      title={mockTitle}
      project_id={mockProject_id}
      color_1_id={mockColor_1_id}
      color_2_id={mockColor_2_id}
      color_3_id={mockColor_3_id}
      color_4_id={mockColor_4_id}
      color_5_id={mockColor_5_id}
    />);

    expect(wrapper).toMatchSnapshot();
  });
});
