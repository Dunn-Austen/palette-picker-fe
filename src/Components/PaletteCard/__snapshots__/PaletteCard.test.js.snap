// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaletteCard should match the snapshot 1`] = `
<section
  className="palette-card"
>
  <section
    className="palette-container"
  >
    <div
      className="palette-slice"
      style={
        Object {
          "backgroundColor": "#bbbd33",
        }
      }
    />
    <div
      className="palette-slice"
      style={
        Object {
          "backgroundColor": "#bbbd33",
        }
      }
    />
    <div
      className="palette-slice"
      style={
        Object {
          "backgroundColor": "#bbbd33",
        }
      }
    />
    <div
      className="palette-slice"
      style={
        Object {
          "backgroundColor": "#bbbd33",
        }
      }
    />
    <div
      className="palette-slice"
      style={
        Object {
          "backgroundColor": "#bbbd33",
        }
      }
    />
  </section>
  <section
    className="btn-container"
  >
    <button
      className="edit-palette"
      onClick={[Function]}
    >
      Edit
    </button>
    <button
      className="delete-palette"
      data-id={1}
      onClick={[Function]}
    >
      Delete
    </button>
  </section>
</section>
`;
